jeeps/gpsmath.cc

```diff
@@ -1516,6 +1517,98 @@ void GPS_Math_Molodensky(double Sphi, double Slam, double SH, double Sa,
+** @param [r] sppm [double] scale factor - 1 (ppm)
+** @param [r] rXs  [double] rotation about x axis (seconds)
+** @param [r] rYs  [double] rotation about y axis (seconds)
+** @param [r] rZs  [double] rotation about z axis (seconds)
+**
+** @return [void]
+************************************************************************/
+void GPS_Math_Helmert(double Sx, double Sy, double Sz,
+                      double* Dx, double* Dy, double* Dz,
+                      double tX, double tY, double tZ,
+                      double sppm,
+                      double rXs, double rYs, double rZs)
+{
+  double s = sppm/1000000.0;
+  double rX = (rXs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+  double rY = (rYs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+  double rZ = (rZs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+
+  // rotation/scaling matrix = [1+s -rz ry;rz 1+s -rx;-ry rx 1+s];
+  *Dx = tX + (1 + s)*Sx + -rZ*Sy + rY*Sz;
+  *Dy = tY + rZ*Sx + (1 + s)*Sy + -rX*Sz;
+  *Dz = tZ + -rY*Sx + rX*Sy + (1 + s)*Sz;
+
+  return;
+}
+
+/* @func GPS_Math_Inverse_Helmert ***********************************
+**
+** Transform from source to coverted datum.
+** translation, scale and rotation parameters are the Helmert parameters
+** to traslate from the converted to source datum!
+**
+** @param [r] Sx [double] source X
+** @param [r] Sy [double] source Y
+** @param [r] Sz [double] source Z
+** @param [w] Dx [double*] converted X
+** @param [w] Dy [double*] converted Y
+** @param [w] Dz [double*] converted Z
+** @param [r] tX   [double] translation along x axis (meters)
+** @param [r] tY   [double] translation along y axis (meters)
+** @param [r] tZ   [double] translation along z axis (meters)
+** @param [r] sppm [double] scale factor - 1 (ppm)
+** @param [r] rXs  [double] rotation about x axis (seconds)
+** @param [r] rYs  [double] rotation about y axis (seconds)
+** @param [r] rZs  [double] rotation about z axis (seconds)
+**
+** @return [void]
+************************************************************************/
+void GPS_Math_Inverse_Helmert(double Sx, double Sy, double Sz,
+                              double* Dx, double* Dy, double* Dz,
+                              double tX, double tY, double tZ,
+                              double sppm,
+                              double rXs, double rYs, double rZs)
+{
+  double s = sppm/1000000.0;
+  double rX = (rXs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+  double rY = (rYs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+  double rZ = (rZs/(60.0*60.0))*GPS_PI/(180.0); /* radians */
+
+  // forward rotation/scaling matrix is [1+s -rz ry;rz 1+s -rx;-ry rx 1+s]
+  // compute inverse of forward Helmert rotation/scaling matrix
+  // https://www.wolframalpha.com/input?i2d=true&i=Power%5B%7B%7B1%2Bs%2C-Subscript%5Br%2Cz%5D%2CSubscript%5Br%2Cy%5D%7D%2C%7BSubscript%5Br%2Cz%5D%2C1%2Bs%2C-Subscript%5Br%2Cx%5D%7D%2C%7B-Subscript%5Br%2Cy%5D%2CSubscript%5Br%2Cx%5D%2C1%2Bs%7D%7D%2C-1%5D
```

Comment:
Within rounding errors GPS_Math_Inverse_Helmert reverses GPS_Math_Helmert.  I coded GPS_Math_Inverse_Helmert when I was developing the test from BNG to WGS84 and back to BNG.  The issue with getting that round trip test to work ended up being elsewhere, so at this point we are just changing the signs of the 7 parameters and using GPS_Math_Helmert to go backwards.  This approximation was mentioned in the ordinance survey reference.  The selection of inversion method in GPS_Math_UKOSMap_To_WGS84_H is done with conditional compilation.  I left GPS_Math_Inverse_Helmert in because it exactly undoes GPS_Math_Inverse_Helmert which might be useful if we have trouble with the approximate method.  I don't have a citation for the exact method, but it is straight forward linear algebra (thanks to WolframAlpha).